BB2-448 Create multi stage PR checks for continuous-integration/jenkins/pr-merge

dtisza1 opened the pull request:
<!--
You've got a Pull Request you want to submit? Awesome!
This PR template is here to help ensure you're setup for success:
  please fill it out to help ensure that your PR is complete and ready for approval.
-->

**JIRA Ticket:**
[BB2-448](https://jira.cms.gov/browse/BB2-448)
[BB2-449](https://jira.cms.gov/browse/BB2-449)


**User Story or Bug Summary:**
<!-- Please copy-paste the brief user story or bug description that this PR is intended to address. -->
For BB2-448, this is to improve our Django Unit Tests and have them run inside the new CB Core.  This will also have the benefit of moving away from using Travis CI and improve the performance & reliability of the checks.

For BB2-449, this is to improve our Python Linting Check for PRs and have them run inside the new CB Core.  This will also have the benefit of moving away from using Travis CI and improve the performance & reliability of the checks.

This combines linting/Flake8, unit testing and integration testing together under a single CBC Jenkins project/job called "CHECK - Github PRs". It runs these via different stages in the multibranch pipeline.

### What Does This PR Do?

<!--
Add detailed description & discussion of changes here.
The contents of this section should be used as your commit message (unless you merge the PR via a merge commit, of course).

Please follow standard Git commit message guidelines:
* First line should be a capitalized, short (50 chars or less) summary.
* The rest of the message should be in standard Markdown format, wrapped to 72 characters.
* Describe your changes in imperative mood, e.g. "make xyzzy do frotz" instead of "[This patch] makes xyzzy do frotz" or "[I] changed xyzzy to do frotz", as if you are giving orders to the codebase to change its behavior.
* List all relevant Jira issue keys, one per line at the end of the message, per: <https://confluence.atlassian.com/jirasoftwarecloud/processing-issues-with-smart-commits-788960027.html>.

Reference: <https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project>.
-->
**The following is a list of changes from the commits (including additional comments):**

* Remove Travis/TOX completely 
  - Remove .travis.yml and tox.ini files.  Travis functionality is also being removed for BB2-449, so it makes sense to remove completely here.  BB2-449 changes are being included in this PR.

* Rename Jenkins and Kubernetes POD config files
  - Rename Kubernetes POD config file from cbc-run-integration-tests.yaml to cbc-pod-deployment-config.yml.
  - Rename POD config file referrence and rename file Jenkinsfile.cbc-run-integration-tests to Jenkinsfile.cbc-run-multi-pr-checks

* Add unit tests stage to pipeline
  - Add unit tests as a new stage in Jenkinsfiles/Jenkinsfile.cbc-run-multi-pr-checks

* Unset ENV vars in runtests.py
  - Unset ENV vars in runtests.py for unit type tests for DJANGO_USER_ID_ITERATIONS' and 'DJANGO_USER_ID_SALT'

* Add flake8 configuration file .flake8

* Add unit tests stage to pipeline
  - Add unit tests as a new stage in Jenkinsfiles/Jenkinsfile.cbc-run-multi-pr-checks

* Update DEV requirements for require-hashes and /vendor/
  - In Makefile, update the "reqs-download" option to download package files for both requirements.txt and requirements.dev.txt in to the ./vendor/ directory.
  - In Makefile, update the "reqs-install" and "reqs-install-dev" options to require checking the hashes via the --require-hashes option.
  - Include extra DEV related packages under ./vendor/ directory.
  - Remove "make reqs-install" since "make reqs-install-dev" will install all needed packages in Jenkinsfiles/Jenkinsfile.cbc-run-multi-pr-checks

* Remove require-hashes PIP install option until after amz linux tests.
  - This conflicts with current amazon linux testing. Will backlog this change for later.


### What Should Reviewers Watch For?

<!--
Add some items to the following list, or remove the entire section if it doesn't apply for some reason.

Common items include:
* Is this likely to address the goals expressed in the user story?
* Are any additional documentation updates needed?
* Are there any unhandled and/or untested edge cases you can think of?
* Is user input properly sanitized & handled?
* Does this make any backwards-incompatible changes that might break end user clients?
* Can you find any bugs if you run the code locally and test it manually?
-->
If you're reviewing this PR, please check these things, in particular:

* Are there any unhandled and/or untested edge cases you can think of?
* Does this make any backwards-incompatible changes that might break end user clients?
* Can you find any bugs if you run the code locally and test it manually?
* Naming in general? Do the namings of variables, methods, classes, and other items make sense? Are they too verbose, or not verbose enough? Any redundancies? Are the comments descriptive enough?
* Are audit logging items properly handled? Is sensitive information either not included or hashed?


### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then here's a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence: N/A.
    * Does this PR add any new software dependencies?  **No**.
    * Does this PR modify or invalidate any of our security controls? **No**.
    * Does this PR store or transmit data that was not stored or transmitted before? **No**.
* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR should not be merged unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons? **No**.


### What Needs to Be Merged and Deployed Before this PR?

<!--
Add some items to the following list, or remove the entire section if it doesn't apply.

Common items include:
* Database migrations (which should always be deployed by themselves, to reduce risk).
* New features in external dependencies (e.g. BFD).
-->


### Submitter Checklist

<!--
Helpful hint: if needed, Git allows you to edit your PR's commits and history, prior to merge.
See these resources for more information:

* <https://dev.to/maxwell_dev/the-git-rebase-introduction-i-wish-id-had>
* <https://raphaelfabeni.com/git-editing-commits-part-1/>
-->

I have gone through and verified that...:

* [ X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [ X] I have named this PR and its branch such that they'll be automatically be linked to the (most) relevant Jira issue, per: <https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html>.
* [X ] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X ] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X ] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X ] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. BFD, SLS, etc.).
* [X ] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

njdister commented on Jenkinsfiles/Jenkinsfile.cbc-run-multi-pr-checks:
```diff
@@ -24,29 +24,47 @@ pipeline {
+      }
+    }
+
+    stage("RUN integration tests") {
```
Silly complaint but you could capitalize words here for Jenkins consistency.  😅

njdister reviewed (approved):
Very simple and straightforward PR that does a lot, awesome work on this!

Not a blocker but what do you think about vendoring our dev dependencies?  Since we've now added a dependency on pulling packages from PyPI to run unit tests, I wonder about performance and availability implications.  Like I said, not a blocker, but perhaps worth a backlog ticket.

njdister commented on Makefile:
```diff
@@ -2,11 +2,12 @@ reqs-compile:
 
 reqs-install:
-	pip install -r requirements/requirements.txt --no-index --find-links ./vendor/
+	pip install --require-hashes -r requirements/requirements.txt --no-index --find-links ./vendor/
```
Unfortunately I think `--require-hashes` will break Amazon Linux 2/Python 3.7 migration if we want to allow side by side deployments without updating the hashes since we now have both versions:

```
- name: "install requirements"
  become_user: "{{ remote_admin_account }}"
  become: yes
  shell: |
    source {{ app_py_virtual_env }}/bin/activate && make reqs-install
  args:
    chdir: "{{ app_pyapp_home }}/{{ common_project_name }}"
  register: out
```

dtisza1 replied:
@njdister I just removed the `--require-hashes` option for now. Will add a backlog Jira ticket for it.

dtisza1 commented:
> Very simple and straightforward PR that does a lot, awesome work on this!
> 
> Not a blocker but what do you think about vendoring our dev dependencies? Since we've now added a dependency on pulling packages from PyPI to run unit tests within Jenkins, I wonder about performance and availability implications. Like I said, not a blocker, but perhaps worth a backlog ticket.

@njdister I just pushed up an update to download the DEV requirements packages to the ./vendor/ directory.  I also added in the --require-hashes option for PIP installs.  This looks to be working well.

JFU-GIT reviewed (approved):
looks good to me, good to see user id salt and iteration env vars cleared for tests.
